### ISSUE TITLE
can't get the real ip when proxying Fly apps with Cloudflare

### ISSUE THREAD

**Issue by luizkowalski** [modosc/cloudflare-rails#72]
hey 👋🏻 

first of all, thanks for the gem, really nice work!
second: I'm not 100% sure that this problem is related to the gem but I figured I might raise this here since you probably know more than I in this topic :D 

I have an app running on [fly.io](https://fly.io/) and when an app is deployed on fly, it gets proxied by Fly's load balancer, therefore `request.remote_ip` doesn't quite work as it will always return the load balancer ip. More about it [here](https://fly.io/docs/reference/runtime-environment/#fly-client-ip)

on top of that, I fully proxied my app with Cloudlfare, meaning that now, `Fly-Client-IP` shouldn't work anymore (at least that's what I understand) because it will be now the Cloudlfare's ip.

When I added this gem and inspected both `ip` and `remote_ip` I get a different ip than mine. This is what I see:

```ruby
class ApplicationController < ActionController::Base
  before_action do
    Rails.logger.info { "Ip1: #{request.remote_ip}" } 
    Rails.logger.info { "Ip2: #{request.ip}" }
    Rails.logger.info { "Ip3: #{request.env['HTTP_CF_CONNECTING_IP']}" }
    Rails.logger.info { "Ip4: #{request.env['HTTP_FLY_CLIENT_IP']}" }
    Rails.logger.info { "Ip5: #{request.env['HTTP_X_FORWARDED_FOR']}" }
    Rails.logger.info { "env: #{request.env}" }
  end
end
```

```bash
2022-12-18T16:29:02Z app[f6619f7d] fra [info][5aafb330-2ea4-4d7a-aabf-75747c91e571] Ip1: 37.16.13.28
2022-12-18T16:29:02Z app[f6619f7d] fra [info][5aafb330-2ea4-4d7a-aabf-75747c91e571] Ip2: 37.16.13.28
2022-12-18T16:29:02Z app[f6619f7d] fra [info][5aafb330-2ea4-4d7a-aabf-75747c91e571] Ip3: 186.194.148.180
2022-12-18T16:29:02Z app[f6619f7d] fra [info][5aafb330-2ea4-4d7a-aabf-75747c91e571] Ip4: 172.70.55.48
2022-12-18T16:29:02Z app[f6619f7d] fra [info][5aafb330-2ea4-4d7a-aabf-75747c91e571] Ip5: 186.194.148.180, 37.16.13.28
```

so the correct IP should be IP n.3. `HTTP_X_FORWARDED_FOR` is also _kinda_ right becuase it includes my IP as well but IIRC, Rails will fetch the last ip in the list.

any idea if there is something that could be done here or should I just rely on `HTTP_CF_CONNECTING_IP` in my case?

thanks!

**Comment by modosc** [comment]
if i'm reading this right, `37.16.13.28` is fly.io, `172.70.55.48` is cloudflare, and `186.194.148.180` is your client address when you make a request. 

in general to parse `X-FORWARDED-FOR` you start at the left-most entry. if it's trusted you can go to the right one address and check again. as soon as you hit an address that's not trusted you use that as your client ip address.

in rails specifically this is detailed here: https://api.rubyonrails.org/v7.0/classes/ActionDispatch/RemoteIp.html

so in the above scenario,  cloudflare should be adding an `X-Forwarded-For: 186.194.148.180, 37.16.13.28` header before the traffic is forwarded off to flyio. it looks like fly.io may be dropping this header or trying to do something clever based on the existing `X-Forwarded-For` / `CF-Connecting-ip` header. either way it's creating an `X-Forwarded-For` header _without_ cloudflare's ip.

when rails sees the request, it looks at the rightmost ip address in `X-Forwarded-For`, doesn't find it in the trusted proxy list, and therefore decides that's the address it should return for `request.ip` / `request.remote_ip`.

this gem works by adding all of the published cloudflare netblocks to rails' trusted proxy list - cloudflare publishes this via an api. in the above case you'd need a list of the address ranges fly.io uses and [they don't appear to have an api setup for this](https://community.fly.io/t/request-fly-ip-ranges/127/24).

tldr: i don't believe this gem will help you. even if you open a ticket with fly.io around the dropped header you'd see the same results since `37.16.13.28` is untrusted.

hope this helps!
